File: SteamAPIWrap/MatchMakingServers.cpp
#include "Precompiled.hpp"
#include "MatchMakingServers.hpp"

using namespace SteamAPIWrap;

namespace SteamAPIWrap
{
	template <> CMatchmakingServers * CSingleton<CMatchmakingServers>::instance = nullptr;


	CMatchmakingServers::CMatchmakingServers()
		: matchmakingServers(nullptr)
	{

	}

	ServerListRequest CMatchmakingServers::RequestInternetServerList(u32 appId, PDataPointer filters,
		u32 filterCount, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestInternetServerList(appId,
			reinterpret_cast<MatchMakingKeyValuePair_t **>(filters), filterCount,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));

		// Handle the pointer as an ServerListRequest (u32) on the managed side.
		return reinterpret_cast<ServerListRequest>(request);
	}

	ServerListRequest CMatchmakingServers::RequestLANServerList(u32 appID, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestLANServerList(appID,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));
		return reinterpret_cast<ServerListRequest>(request);
	}

	ServerListRequest CMatchmakingServers::RequestFriendsServerList(u32 appId, PDataPointer filters,
		u32 filterCount, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestFriendsServerList(appId,
			reinterpret_cast<MatchMakingKeyValuePair_t **>(filters), filterCount,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));

		return reinterpret_cast<ServerListRequest>(request);
	}

	ServerListRequest CMatchmakingServers::RequestFavoritesServerList(u32 appId, PDataPointer filters,
		u32 filterCount, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestFavoritesServerList(appId,
			reinterpret_cast<MatchMakingKeyValuePair_t **>(filters), filterCount,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));

		return reinterpret_cast<ServerListRequest>(request);
	}

	ServerListRequest CMatchmakingServers::RequestHistoryServerList(u32 appId, PDataPointer filters,
		u32 filterCount, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestHistoryServerList(appId,
			reinterpret_cast<MatchMakingKeyValuePair_t **>(filters), filterCount,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));

		return reinterpret_cast<ServerListRequest>(request);
	}

	ServerListRequest CMatchmakingServers::RequestSpectatorServerList(u32 appId, PDataPointer filters,
		u32 filterCount, uptr requestServersResponse)
	{
		HServerListRequest request = matchmakingServers->RequestSpectatorServerList(appId,
			reinterpret_cast<MatchMakingKeyValuePair_t **>(filters), filterCount,
			reinterpret_cast<ISteamMatchmakingServerListResponse *>(requestServersResponse));

		return reinterpret_cast<ServerListRequest>(request);
	}

	void CMatchmakingServers::ReleaseRequest(ServerListRequest request)
	{
		matchmakingServers->ReleaseRequest(reinterpret_cast<HServerListRequest>(request));
	}

	PDataPointer CMatchmakingServers::GetServerDetails(ServerListRequest request, int server)
	{
		return matchmakingServers->GetServerDetails(reinterpret_cast<HServerListRequest>(request), server);
	}

	void CMatchmakingServers::CancelQuery(ServerListRequest request)
	{
		matchmakingServers->CancelQuery(reinterpret_cast<HServerListRequest>(request));
	}

	void CMatchmakingServers::RefreshQuery(ServerListRequest request)
	{
		matchmakingServers->RefreshQuery(reinterpret_cast<HServerListRequest>(request));
	}

	bool CMatchmakingServers::IsRefreshing(ServerListRequest request)
	{
		return matchmakingServers->IsRefreshing(reinterpret_cast<HServerListRequest>(request));
	}

	s32 CMatchmakingServers::GetServerCount(ServerListRequest request)
	{
		return matchmakingServers->GetServerCount(reinterpret_cast<HServerListRequest>(request));
	}

	void CMatchmakingServers::RefreshServer(ServerListRequest request, s32 server)
	{
		matchmakingServers->RefreshServer(reinterpret_cast<HServerListRequest>(request), server);
	}

	ServerQuery CMatchmakingServers::PingServer(u32 ip, u16 port, uptr requestServersResponse)
	{
		return matchmakingServers->PingServer(ip, port,
			reinterpret_cast<ISteamMatchmakingPingResponse *>(requestServersResponse));
	}

	ServerQuery CMatchmakingServers::PlayerDetails(u32 ip, u16 port, uptr requestServersResponse)
	{
		return matchmakingServers->PlayerDetails(ip, port,
			reinterpret_cast<ISteamMatchmakingPlayersResponse *>(requestServersResponse));
	}

	ServerQuery CMatchmakingServers::ServerRules(u32 ip, u16 port, uptr requestServersResponse)
	{
		return matchmakingServers->ServerRules(ip, port,
			reinterpret_cast<ISteamMatchmakingRulesResponse *>(requestServersResponse));
	}

	void CMatchmakingServers::CancelServerQuery(s32 serverQuery)
	{
		matchmakingServers->CancelServerQuery(serverQuery);
	}
}

ManagedSteam_API PConstantAnsiString MatchmakingServerNetworkAddress_GetConnectionString(PConstantDataPointer instance)
{
	const servernetadr_t *fakedInstance = reinterpret_cast<const servernetadr_t *>(instance);
	return fakedInstance->GetConnectionAddressString();
}

ManagedSteam_API PConstantAnsiString MatchmakingServerNetworkAddress_GetQueryString(PConstantDataPointer instance)
{
	const servernetadr_t *fakedInstance = reinterpret_cast<const servernetadr_t *>(instance);
	return fakedInstance->GetQueryAddressString();
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestInternetServerList(u32 appId,
	PDataPointer filters, u32 filterCount, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestInternetServerList(appId, filters, filterCount,
		requestServersResponse);
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestLANServerList(u32 appId, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestLANServerList(appId, requestServersResponse);
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestFriendsServerList(u32 appId,
	PDataPointer filters, u32 filterCount, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestFriendsServerList(appId, filters, filterCount,requestServersResponse);
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestFavoritesServerList(u32 appId,
	PDataPointer filters, u32 filterCount, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestFavoritesServerList(appId, filters, filterCount,requestServersResponse);
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestHistoryServerList(u32 appId,
	PDataPointer filters, u32 filterCount, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestHistoryServerList(appId, filters, filterCount,requestServersResponse);
}

ManagedSteam_API ServerListRequest MatchmakingServers_RequestSpectatorServerList(u32 appId,
	PDataPointer filters, u32 filterCount, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().RequestSpectatorServerList(appId, filters, filterCount,requestServersResponse);
}

ManagedSteam_API void MatchmakingServers_ReleaseRequest(ServerListRequest request)
{
	CMatchmakingServers::Instance().ReleaseRequest(request);
}

ManagedSteam_API PDataPointer MatchmakingServers_GetServerDetails(ServerListRequest request, s32 server)
{
	return CMatchmakingServers::Instance().GetServerDetails(request, server);
}

ManagedSteam_API s32 MatchmakingServers_GetGameServerItemSize()
{
	return sizeof(gameserveritem_t);
}

ManagedSteam_API void MatchmakingServers_CancelQuery(ServerListRequest request)
{
	CMatchmakingServers::Instance().CancelQuery(request);
}

ManagedSteam_API void MatchmakingServers_RefreshQuery(ServerListRequest request)
{
	CMatchmakingServers::Instance().RefreshQuery(request);
}

ManagedSteam_API bool MatchmakingServers_IsRefreshing(ServerListRequest request)
{
	return CMatchmakingServers::Instance().IsRefreshing(request);
}

ManagedSteam_API int MatchmakingServers_GetServerCount(ServerListRequest request)
{
	return CMatchmakingServers::Instance().GetServerCount(request);
}

ManagedSteam_API void MatchmakingServers_RefreshServer(ServerListRequest request, s32 server)
{
	CMatchmakingServers::Instance().RefreshServer(request, server);
}

ManagedSteam_API ServerQuery MatchmakingServers_PingServer(u32 ip, u16 port, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().PingServer(ip, port, requestServersResponse);
}

ManagedSteam_API ServerQuery MatchmakingServers_PlayerDetails(u32 ip, u16 port, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().PlayerDetails(ip, port, requestServersResponse);
}

ManagedSteam_API ServerQuery MatchmakingServers_ServerRules(u32 ip, u16 port, uptr requestServersResponse)
{
	return CMatchmakingServers::Instance().ServerRules(ip, port, requestServersResponse);
}

ManagedSteam_API void MatchmakingServers_CancelServerQuery(s32 serverQuery)
{
	CMatchmakingServers::Instance().CancelServerQuery(serverQuery);
}

File: ManagedSteam/Implementations/MatchmakingServers.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Runtime.InteropServices;

using ManagedSteam.SteamTypes;
using ManagedSteam.Utility;

namespace ManagedSteam.Implementations
{
    class MatchmakingServers : SteamService, IMatchmakingServers
    {
        private List<Exception> bufferedExceptions;

        internal MatchmakingServers()
        {

            bufferedExceptions = new List<Exception>();
            ServerListResponse = new Dictionary<uint, MatchmakingServerListResponse>();
            PingResponse = new Dictionary<uint, MatchmakingPingResponse>();
            PlayersResponse = new Dictionary<uint, MatchmakingPlayersResponse>();
            RulesResponse = new Dictionary<uint, MatchmakingRulesResponse>();

            MatchmakingServerListResponse.Initialize();
            MatchmakingPingResponse.Initialize();
            MatchmakingPlayersResponse.Initialize();
            MatchmakingRulesResponse.Initialize();
        }

        internal Dictionary<uint, MatchmakingServerListResponse> ServerListResponse { get; private set; }
        internal Dictionary<uint, MatchmakingPingResponse> PingResponse { get; private set; }
        internal Dictionary<uint, MatchmakingPlayersResponse> PlayersResponse { get; private set; }
        internal Dictionary<uint, MatchmakingRulesResponse> RulesResponse { get; private set; }

        internal override void CheckIfUsableInternal()
        {

        }

        internal override void ReleaseManagedResources()
        {
            foreach (var item in ServerListResponse)
            {
                item.Value.LibraryShuttingDown();
            }

            foreach (var item in PingResponse)
            {
                item.Value.LibraryShuttingDown();
            }

            foreach (var item in PlayersResponse)
            {
                item.Value.LibraryShuttingDown();
            }

            foreach (var item in RulesResponse)
            {
                item.Value.LibraryShuttingDown();
            }
        }

        internal override void InvokeEvents()
        {
            foreach (var item in ServerListResponse)
            {
                item.Value.InvokeEvents();
            }

            foreach (var item in PingResponse)
            {
                item.Value.InvokeEvents();
            }

            foreach (var item in PlayersResponse)
            {
                item.Value.InvokeEvents();
            }

            foreach (var item in RulesResponse)
            {
                item.Value.InvokeEvents();
            }
        }

        internal void ReportExceptions()
        {
            foreach (var item in bufferedExceptions)
            {
                Steam.Instance.ReportException(item);
            }
            bufferedExceptions.Clear();
        }

        internal void SaveException(Exception exception)
        {
            bufferedExceptions.Add(exception);
        }

        private ServerListRequestHandle ServerRequest(ServerRequestType requestType, AppID appID,
            MatchMakingKeyValuePair[] filters, MatchmakingServerListResponse requestResponse)
        {
            // Code reuse FTW!

            // Allocate an array that will hold addresses to native MatchMakingKeyValuePair objects
            using (NativeBuffer arrayBuffer = new NativeBuffer(Marshal.SizeOf(typeof(IntPtr)) * filters.Length))
            {
                NativeBuffer[] nativeObjects = new NativeBuffer[filters.Length];
                // Fill each array slot with an address to an actual object
                try
                {
                    for (int i = 0; i < filters.Length; i++)
                    {
                        // Copy the managed objects to native memory
                        NativeBuffer buffer = NativeBuffer.CopyToNative(filters[i]);
                        nativeObjects[i] = buffer;
                        // Add the native address to the array
                        Marshal.WriteInt32(arrayBuffer.UnmanagedMemory, i * Marshal.SizeOf(typeof(int)), buffer.UnmanagedMemory.ToInt32());
                    }

                    // Now do the actual request

                    switch (requestType)
                    {
                        case ServerRequestType.Internet:
                            return new ServerListRequestHandle(
                                NativeMethods.MatchmakingServers_RequestInternetServerList(appID.AsUInt32,
                                arrayBuffer.UnmanagedMemory, (uint)filters.Length, requestResponse.ObjectId));
                        case ServerRequestType.Friends:
                            return new ServerListRequestHandle(
                                NativeMethods.MatchmakingServers_RequestFriendsServerList(appID.AsUInt32,
                                arrayBuffer.UnmanagedMemory, (uint)filters.Length, requestResponse.ObjectId));
                        case ServerRequestType.Favorites:
                            return new ServerListRequestHandle(
                                NativeMethods.MatchmakingServers_RequestFavoritesServerList(appID.AsUInt32,
                                arrayBuffer.UnmanagedMemory, (uint)filters.Length, requestResponse.ObjectId));
                        case ServerRequestType.History:
                            return new ServerListRequestHandle(
                                NativeMethods.MatchmakingServers_RequestHistoryServerList(appID.AsUInt32,
                                arrayBuffer.UnmanagedMemory, (uint)filters.Length, requestResponse.ObjectId));
                        case ServerRequestType.Spectator:
                            return new ServerListRequestHandle(
                                NativeMethods.MatchmakingServers_RequestSpectatorServerList(appID.AsUInt32,
                                arrayBuffer.UnmanagedMemory, (uint)filters.Length, requestResponse.ObjectId));
                        default:
                            // This should never happen as our code can not be allowed to make this error
                            throw new ArgumentException();
                    }
                }
                finally
                {
                    // Cleanup in all cases. Exception or not.
                    foreach (var item in nativeObjects)
                    {
                        if (item != null)
                        {
                            item.Dispose();
                        }
                    }
                }
            }
        }

        public ServerListRequestHandle RequestInternetServerList(AppID appID,
            MatchMakingKeyValuePair[] filters, MatchmakingServerListResponse requestServersResponse)
        {
            return ServerRequest(ServerRequestType.Internet, appID, filters, requestServersResponse);
        }

        public ServerListRequestHandle RequestLANServerList(AppID appID,
            MatchmakingServerListResponse requestServersResponse)
        {
            return new ServerListRequestHandle(NativeMethods.MatchmakingServers_RequestLANServerList(
                appID.AsUInt32, requestServersResponse.ObjectId));
        }

        public ServerListRequestHandle RequestFriendsServerList(AppID appID, MatchMakingKeyValuePair[] filters,
            MatchmakingServerListResponse requestServersResponse)
        {
            return ServerRequest(ServerRequestType.Friends, appID, filters, requestServersResponse);
        }

        public ServerListRequestHandle RequestFavoritesServerList(AppID appID, MatchMakingKeyValuePair[] filters,
            MatchmakingServerListResponse requestServersResponse)
        {
            return ServerRequest(ServerRequestType.Favorites, appID, filters, requestServersResponse);
        }

        public ServerListRequestHandle RequestHistoryServerList(AppID appID, MatchMakingKeyValuePair[] filters,
            MatchmakingServerListResponse requestServersResponse)
        {
            return ServerRequest(ServerRequestType.History, appID, filters, requestServersResponse);
        }

        public ServerListRequestHandle RequestSpectatorServerList(AppID appID, MatchMakingKeyValuePair[] filters,
            MatchmakingServerListResponse requestServersResponse)
        {
            return ServerRequest(ServerRequestType.Spectator, appID, filters, requestServersResponse);
        }

        public void ReleaseRequest(ServerListRequestHandle serverListRequest)
        {
            NativeMethods.MatchmakingServers_ReleaseRequest(serverListRequest.AsUInt32);
        }

        public GameServerItem GetServerDetails(ServerListRequestHandle request, int server)
        {
            IntPtr returnValue;
            returnValue = NativeMethods.MatchmakingServers_GetServerDetails(request.AsUInt32, server);
            return GameServerItem.Create(returnValue, NativeMethods.MatchmakingServers_GetGameServerItemSize());
        }

        public void CancelQuery(ServerListRequestHandle request)
        {
            NativeMethods.MatchmakingServers_CancelQuery(request.AsUInt32);
        }

        public void RefreshQuery(ServerListRequestHandle request)
        {
            NativeMethods.MatchmakingServers_RefreshQuery(request.AsUInt32);
        }

        public bool IsRefreshing(ServerListRequestHandle request)
        {
            return NativeMethods.MatchmakingServers_IsRefreshing(request.AsUInt32);
        }

        public int GetServerCount(ServerListRequestHandle request)
        {
            return NativeMethods.MatchmakingServers_GetServerCount(request.AsUInt32);
        }

        public void RefreshServer(ServerListRequestHandle request, int server)
        {
            NativeMethods.MatchmakingServers_RefreshServer(request.AsUInt32, server);
        }

        public ServerQueryHandle PingServer(uint ip, ushort port, MatchmakingPingResponse requestServersResponse)
        {
            return new ServerQueryHandle(NativeMethods.MatchmakingServers_PingServer(ip, port, requestServersResponse.ObjectId));
        }
        public ServerQueryHandle PlayerDetails(uint ip, ushort port, MatchmakingPlayersResponse requestServersResponse)
        {
            return new ServerQueryHandle(NativeMethods.MatchmakingServers_PlayerDetails(ip, port, requestServersResponse.ObjectId));
        }
        public ServerQueryHandle ServerRules(uint ip, ushort port, MatchmakingRulesResponse requestServersResponse)
        {
            return new ServerQueryHandle(NativeMethods.MatchmakingServers_ServerRules(ip, port, requestServersResponse.ObjectId));
        }
        public void CancelServerQuery(ServerQueryHandle serverQuery)
        {
            NativeMethods.MatchmakingServers_CancelServerQuery(serverQuery.AsInt32);
        }

        private enum ServerRequestType
        {
            Internet,
            Friends,
            Favorites,
            History,
            Spectator,
        }
    }
}
